internal/component/loki/source/kubernetes/kubetail/tailer_test.go

```diff
@@ -42,3 +53,78 @@ func Test_parseKubernetesLog(t *testing.T) {
 		})
 	}
 }
+
+// Test context cancellation.
+func TestContextCancel(t *testing.T) {
+	logger := log.NewLogfmtLogger(os.Stdout)
+
+	target := NewTarget(
+		labels.FromStrings("t1", "t1", "t2", "t2"),
+		labels.FromStrings("p1", "p1", "p2", "p2"),
+	)
+
+	pos, err := positions.New(logger, positions.Config{
+		SyncPeriod:    10 * time.Second,
+		PositionsFile: t.TempDir() + "/positions.yml",
+	})
+	require.NoError(t, err)
+
+	namespace := "alloy-test-ns"
+	podName := "alloy-pod1"
+	kubeObjects := []runtime.Object{
+		&corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      podName,
+				Namespace: namespace,
+				Labels: map[string]string{
+					"label1": "value1",
+				},
+			},
+		},
+	}
+	fakeClientset := fake.NewClientset(kubeObjects...)
+	entries := make(chan loki.Entry)
+	handler := loki.NewEntryHandler(entries, func() {})
+
+	task := &tailerTask{
+		Options: &Options{
+			Client:    fakeClientset,
+			Handler:   handler,
+			Positions: pos,
+		},
+		Target: target,
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	tailer := newTailer(logger, task)
+	go tailer.Run(ctx)
+
+	// Let the tailer run for a bit.
+	time.Sleep(10 * time.Second)
+
+	// deletionOption := metav1.NewDeleteOptions(0)
+	// err = fakeClientset.CoreV1().Pods(namespace).Delete(ctx, podName, *deletionOption)
+	// require.NoError(t, err)
+
+	// fakeClientset.CoreV1().(*fakecorev1.FakeCoreV1).PrependReactor("get", "pods", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+	// 	return true, &v1.Pod{}, errors.New("Error creating secret")
+	// })
+
+	// fakeClientset.CoreV1().(*fakecorev1.FakeCoreV1).PrependReactor("logs", "alloy-pod1", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
```

Comment:
The main problem right now is that I can't get this line to work. I'm not sure how to fake a k8s API which doesn't send any log lines.